File: client/src/lib/receive-quality.ts
import { useSyncExternalStore } from "react";

/**
 * The largest picture this device asks the SFU for, remembered per device.
 *
 * WHY THIS EXISTS. A 100-viewer watch party on 5 Sep 2026 cost 323 GB of SFU
 * downstream in three and a half hours, and most of those viewers were on
 * phones drawing a 1080p share into a 390-pixel-wide element. The presenter
 * now publishes simulcast layers (`video-quality.ts`), so for the first time
 * there is a smaller copy on the server to ask for, and this is the asking.
 *
 * PURELY LOCAL, on purpose, the same call `participant-rail-preference` makes:
 * "this is a phone on mobile data" describes a device, not a person, and the
 * same account on a desktop should get the desktop default without the phone
 * having to be undone first.
 *
 * AUTO IS ADAPTIVE, NOT "HIGH". With `adaptiveStream` on the room, LiveKit
 * measures the element the video is drawn into and asks the SFU for the
 * smallest layer that covers it. A fixed choice is a ceiling laid over that:
 * `RemoteTrackPublication.setVideoQuality` names the highest layer this side
 * will accept, and the adaptive measurement still shrinks below it when the
 * element is small. See `livekit-session.ts` for the library-version note.
 *
 * MESH IGNORES THIS ENTIRELY. A mesh presenter encodes one stream per peer and
 * `RTCRtpReceiver` has no size parameter, so on that transport the menu hides
 * the selector and keeps the sentence that says the sender chooses.
 */

export const RECEIVE_QUALITIES = ["auto", "1080p", "720p", "360p"] as const;

export type ReceiveQuality = (typeof RECEIVE_QUALITIES)[number];

const STORAGE_KEY = "pqp:receive-quality";

/** Storage hands back `unknown`; this is the only door in. */
export function parseReceiveQuality(raw: unknown): ReceiveQuality | null {
  return RECEIVE_QUALITIES.includes(raw as ReceiveQuality)
    ? (raw as ReceiveQuality)
    : null;
}

/** What decides a device's default before anybody has opened the menu. */
export interface ReceiveDeviceSignals {
  /** `(pointer: coarse)`: a finger, so a phone or a tablet. */
  coarsePointer: boolean;
  /** The viewport is narrower than a laptop, whatever the pointer. */
  smallViewport: boolean;
  /** The user agent says iPhone, iPad or Android. */
  mobileUserAgent: boolean;
  /**
   * The link is metered or slow: the Network Information API says cellular,
   * an effective type of 3G or worse, or the user turned on data saving.
   * Optional because the callers that predate it pass the three above.
   */
  cellular?: boolean;
}

/** Below this, a window is a phone or a tablet in portrait, not a desktop. */
const SMALL_VIEWPORT_MAX_PX = 900;

/** `navigator.connection` as far as this file reads it. Chromium and Android only. */
interface NetworkInformationLike {
  type?: string;
  effectiveType?: string;
  saveData?: boolean;
  addEventListener?: (type: "change", listener: () => void) => void;
  removeEventListener?: (type: "change", listener: () => void) => void;
}

function networkInformation(): NetworkInformationLike | null {
  try {
    if (typeof navigator === "undefined") {
      return null;
    }
    const nav = navigator as Navigator & {
      connection?: NetworkInformationLike;
      mozConnection?: NetworkInformationLike;
      webkitConnection?: NetworkInformationLike;
    };
    return nav.connection ?? nav.mozConnection ?? nav.webkitConnection ?? null;
  } catch {
    return null;
  }
}

const SLOW_EFFECTIVE_TYPES = new Set(["slow-2g", "2g", "3g"]);

/** The cellular signal, from the shape the API hands over. Pure. */
export function isCellularConnection(
  connection: NetworkInformationLike | null | undefined,
): boolean {
  if (!connection) {
    return false;
  }
  return (
    connection.type === "cellular" ||
    SLOW_EFFECTIVE_TYPES.has(connection.effectiveType ?? "") ||
    connection.saveData === true
  );
}

/**
 * Read the four signals off the browser. Each one is wrapped because none
 * exists in a node test, in an old WebView, or in some Electron partitions.
 * Safari and Firefox have no `navigator.connection`, so on those the
 * cellular signal is simply false and the device default stands.
 */
export function readReceiveDeviceSignals(): ReceiveDeviceSignals {
  let coarsePointer = false;
  let smallViewport = false;
  let mobileUserAgent = false;
  let cellular = false;
  try {
    coarsePointer =
      typeof window !== "undefined" &&
      typeof window.matchMedia === "function" &&
      window.matchMedia("(pointer: coarse)").matches;
  } catch {
    coarsePointer = false;
  }
  try {
    smallViewport =
      typeof window !== "undefined" &&
      typeof window.innerWidth === "number" &&
      window.innerWidth > 0 &&
      window.innerWidth < SMALL_VIEWPORT_MAX_PX;
  } catch {
    smallViewport = false;
  }
  try {
    mobileUserAgent =
      typeof navigator !== "undefined" &&
      /iPhone|iPad|iPod|Android/i.test(navigator.userAgent ?? "");
  } catch {
    mobileUserAgent = false;
  }
  try {
    cellular = isCellularConnection(networkInformation());
  } catch {
    cellular = false;
  }
  return { coarsePointer, smallViewport, mobileUserAgent, cellular };
}

/**
 * Phones and tablets start at 720p; desktops start on auto; anything on
 * mobile data starts at 360p.
 *
 * 720p rather than 360p because a phone held sideways is a 720-line screen,
 * and the whole point of auto is that the element decides: a phone on auto
 * would already land near 720p in landscape and near 360p in portrait. The
 * fixed 720p is a *ceiling* for the device on top of that, so a tablet drawn
 * at full width never asks for the 1080p layer by accident. A desktop has the
 * pixels and usually the link, so it gets the adaptive default and can pick
 * 1080p by name. Anyone can pick 1080p.
 *
 * CELLULAR WINS OVER THE OTHER THREE. A metered link is a fact about the
 * bill, not the screen: a phone on 5G still pays per byte, and a laptop
 * tethered to one pays the same. 360p of a share is readable text at phone
 * size and a quarter of the 720p layer's bytes. It is a default, so the
 * person who wants 720p on the bus picks it once and it stays picked.
 */
export function defaultReceiveQuality(
  signals: ReceiveDeviceSignals,
): ReceiveQuality {
  if (signals.cellular) {
    return "360p";
  }
  return signals.coarsePointer || signals.smallViewport || signals.mobileUserAgent
    ? "720p"
    : "auto";
}

/** The choice remembered on this device, if the user ever made one. */
export function readStoredReceiveQuality(): ReceiveQuality | null {
  try {
    return parseReceiveQuality(localStorage.getItem(STORAGE_KEY));
  } catch {
    // Storage denied (privacy mode, an Electron partition without quota):
    // the device default is the app working.
    return null;
  }
}

export function loadReceiveQuality(
  signals: ReceiveDeviceSignals = readReceiveDeviceSignals(),
): ReceiveQuality {
  return readStoredReceiveQuality() ?? defaultReceiveQuality(signals);
}

/**
 * Why the standing quality is what it is, when it was not the user's pick.
 * `null` means either the user chose it or it is the plain device default.
 * The menu turns `"cellular"` into one sentence, because a phone that opens
 * the menu on 360p with no memory of choosing it reads as a broken setting.
 */
export type ReceiveQualityReason = "cellular" | null;

/**
 * The reason for a default, given the signals. Pure, so the store and the
 * tests agree on it.
 */
export function receiveQualityReason(
  signals: ReceiveDeviceSignals,
  explicit: boolean,
): ReceiveQualityReason {
  if (explicit) {
    return null;
  }
  return signals.cellular ? "cellular" : null;
}

export function saveReceiveQuality(quality: ReceiveQuality): void {
  try {
    localStorage.setItem(STORAGE_KEY, quality);
  } catch {
    // The choice still applies for the session; only the memory of it is lost.
  }
}

// -------------------------------------------------------------------- store

/**
 * One in-memory copy, shared by the menu that sets it and the session that
 * applies it, so neither has to be threaded through the call stage's props.
 * Lazily read so a test can stub storage before the first read.
 */
let current: ReceiveQuality | null = null;
/** True once a stored choice was read or the menu was used. Never unset. */
let explicit = false;
let reason: ReceiveQualityReason = null;
const listeners = new Set<(quality: ReceiveQuality) => void>();

/**
 * The connection watcher, started on the first read. A phone that leaves
 * Wi-Fi mid-call fires `change` on `navigator.connection`; if the user never
 * picked a size, the default is recomputed and the session follows it like
 * any other change. A picked size is never touched: the whole promise of the
 * menu is that a choice stays chosen.
 */
let watching: NetworkInformationLike | null = null;
function onConnectionChange() {
  const signals = readReceiveDeviceSignals();
  reason = receiveQualityReason(signals, explicit);
  if (explicit) {
    return;
  }
  const next = defaultReceiveQuality(signals);
  if (next === current) {
    return;
  }
  current = next;
  for (const listener of listeners) {
    listener(next);
  }
}

function watchConnection(): void {
  if (watching) {
    return;
  }
  const connection = networkInformation();
  if (!connection || typeof connection.addEventListener !== "function") {
    return;
  }
  try {
    connection.addEventListener("change", onConnectionChange);
    watching = connection;
  } catch {
    // An API that exists but refuses listeners: the first read still counts.
  }
}

function ensureLoaded(): ReceiveQuality {
  if (current === null) {
    const stored = readStoredReceiveQuality();
    const signals = readReceiveDeviceSignals();
    explicit = stored !== null;
    current = stored ?? defaultReceiveQuality(signals);
    reason = receiveQualityReason(signals, explicit);
    watchConnection();
  }
  return current;
}

export function getReceiveQuality(): ReceiveQuality {
  return ensureLoaded();
}

/** Why the standing quality is a default rather than a choice. See the type. */
export function getReceiveQualityReason(): ReceiveQualityReason {
  ensureLoaded();
  return reason;
}

export function setReceiveQuality(quality: ReceiveQuality): void {
  ensureLoaded();
  // Choosing the size the default already was is still a choice: the reason
  // line goes away and a connection change no longer moves it. Choosing the
  // size already chosen is nothing, and nobody is told.
  const changed = quality !== current || reason !== null;
  explicit = true;
  reason = null;
  current = quality;
  saveReceiveQuality(quality);
  if (!changed) {
    return;
  }
  for (const listener of listeners) {
    listener(quality);
  }
}

export function subscribeReceiveQuality(
  listener: (quality: ReceiveQuality) => void,
): () => void {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

/** Test seam: forget the in-memory copy so the next read hits storage again. */
export function resetReceiveQualityForTests(): void {
  current = null;
  explicit = false;
  reason = null;
  listeners.clear();
  if (watching) {
    try {
      watching.removeEventListener?.("change", onConnectionChange);
    } catch {
      // Gone already.
    }
    watching = null;
  }
}

export function useReceiveQuality(): ReceiveQuality {
  return useSyncExternalStore(subscribeReceiveQuality, getReceiveQuality);
}

export function useReceiveQualityReason(): ReceiveQualityReason {
  return useSyncExternalStore(subscribeReceiveQuality, getReceiveQualityReason);
}

File: android/app/src/main/kotlin/gg/pqp/app/voice/ReceiveQuality.kt
package gg.pqp.app.voice

import io.livekit.android.room.track.Track
import io.livekit.android.room.track.VideoQuality

/**
 * Whether this client asks the SFU for a publication at all.
 *
 * The room is joined with `autoSubscribe = false` precisely so this decision
 * exists, and it is a top-level function rather than a `when` inside the
 * engine because it is the only part of the subscription path a test on this
 * machine can reach: nothing in that module can build a `Room` without a
 * device.
 *
 * **Audio, from any source.** A microphone is a voice and a
 * `SCREEN_SHARE_AUDIO` publication is a watch party's soundtrack; both are
 * small and both are wanted. The engine tells them apart afterwards: the
 * presentation's sound is silenced by deafen like everything else and is never
 * what makes somebody count as audible.
 *
 * **Video when it is a screen share, and now when it is a camera.** The share
 * is what a phone joins a LiveKit room to watch, and the presenter publishes
 * simulcast layers (`client/src/lib/video-quality.ts`), so there is a
 * phone-sized copy on the server to ask for; [screenReceiveLayerFor] names it.
 * A camera used to be refused here, because this client drew no camera tiles
 * and a subscribed camera could only be decoded and dropped. It draws them now
 * (the strip under the call bar and the viewer behind it), so refusing one
 * would mean a room where every face is missing on Android alone, which is what
 * the 5 Sep watch party looked like from a phone.
 *
 * Admitting a camera is only half of not paying for it: what keeps twenty of
 * them off a phone's bill is [CameraDemand], which pauses every camera no
 * surface is drawing, and [cameraReceiveLayerFor], which never asks for the top
 * layer. Subscribing is cheap; delivery is what costs, and delivery is off by
 * default.
 *
 * Anything LiveKit adds later is still refused by default, for the reason the
 * camera no longer is: there would be nothing on this client able to draw it.
 */
fun livekitSubscribesTo(kind: Track.Kind, source: Track.Source): Boolean = when (kind) {
    Track.Kind.AUDIO -> true
    Track.Kind.VIDEO -> source == Track.Source.SCREEN_SHARE || source == Track.Source.CAMERA
    else -> false
}

/**
 * The largest simulcast layer this phone asks the SFU for, before anybody has
 * touched a setting.
 *
 * The web's `receive-quality.ts` decides the same thing from four browser
 * signals; on a phone three of them are always true, so what is left is the
 * link. A 100-viewer watch party on 5 Sep 2026 cost 323 GB of SFU downstream
 * in three and a half hours, and most of those viewers were phones drawing a
 * 1080p share into a 390-pixel-wide element.
 *
 * **720p (`MEDIUM`) on Wi-Fi.** A phone held sideways is a 720-line screen,
 * and adaptive stream still shrinks below this ceiling when the picture is
 * drawn smaller. The 1080p layer is never asked for by default: a tablet drawn
 * at full width would take it by accident and nobody would see the difference.
 *
 * **360p (`LOW`) on a metered link.** Mobile data is a fact about the bill,
 * not the screen: 360p of a share is readable text at phone size and a quarter
 * of the 720p layer's bytes. `ConnectivityManager.isActiveNetworkMetered` is
 * the signal, which also covers a hotspot the OS knows is metered.
 *
 * A ceiling, not a request: with adaptive stream on, LiveKit sends the smaller
 * of this and whatever layer covers the view (verified on the web against
 * livekit-client 2.21.0, and the Android SDK's `RemoteTrackPublication` carries
 * the same `videoQuality` and `videoDimensions` pair to the signal). `HIGH` is
 * the SDK's resting value, so this never returns it: returning it would mean
 * "no ceiling", which is the desktop default and not a phone's.
 */
fun screenReceiveLayerFor(metered: Boolean): VideoQuality =
    if (metered) VideoQuality.LOW else VideoQuality.MEDIUM

/**
 * The layer this phone asks for of somebody's **camera**, given how big the
 * surface drawing it is.
 *
 * A separate rule from [screenReceiveLayerFor] because the two are not the same
 * picture. A share is text somebody has to read, drawn one at a time and full
 * bleed; a camera is a face, and the common case is several of them at once in
 * a strip about a hundred dp wide. Asking for 720p of a face to draw it at
 * tile size is the exact waste the share rule was written to stop, multiplied
 * by however many people have their camera on.
 *
 * - **A tile takes the bottom layer**, on any link. At the rail's size that is
 *   more pixels than the tile has, and a room where six people are on camera
 *   costs a phone about what one share does.
 * - **The viewer takes 360p (`MEDIUM`) on Wi-Fi**, because it is the whole
 *   screen and a face at the bottom layer is visibly soft that big.
 * - **A metered link stays at the bottom layer even full screen.** Same
 *   argument as the share rule: mobile data is a fact about the bill.
 *
 * `HIGH` is never returned, for the same reason it never is for a share.
 *
 * **What this is worth today, stated honestly.** The web publishes its camera
 * with `simulcast: false` (`client/src/lib/livekit-session.ts`), so for a web
 * participant there is exactly one layer on the server and this ceiling has
 * nothing smaller to pick: the bytes saved for that publisher come entirely
 * from pausing cameras nobody is drawing. iOS publishes through
 * `setCamera`, which does simulcast, and a camera published with layers is
 * what this rule is already correct for. It is applied to every camera either
 * way, because the alternative is a rule that has to be remembered the day
 * simulcast is switched on.
 */
fun cameraReceiveLayerFor(metered: Boolean, surface: CameraSurface): VideoQuality = when {
    metered -> VideoQuality.LOW
    surface == CameraSurface.Fullscreen -> VideoQuality.MEDIUM
    else -> VideoQuality.LOW
}

/**
 * What to tell the SFU about one camera: whether to send it, and how big.
 *
 * A null [quality] means "do not say anything about the layer", which is not
 * the same as a low one: a publication being paused needs one message, and
 * naming a layer for a stream that is about to stop is a second message that
 * changes nothing.
 */
data class CameraDelivery(val enabled: Boolean, val quality: VideoQuality?)

/**
 * The whole camera delivery decision, in one pure place.
 *
 * It lives here rather than inside [LiveKitEngine] for the reason everything
 * else in this file does: nothing in that module can build a `Room` without a
 * device, so a rule left inside it is a rule nothing on this machine can check.
 * The engine's job is reduced to holding the lock, reading the three inputs and
 * making two SDK calls.
 *
 * - **No surface means paused**, which is the resting state of every camera in
 *   a room this phone has not scrolled to.
 * - **A muted camera is never delivered**, whatever any surface wants. There
 *   are no frames at the far end to send; asking for them is a subscription
 *   paying for keepalives, and the tile has been taken down anyway.
 * - Otherwise the layer is [cameraReceiveLayerFor] of the largest surface
 *   drawing it, sent *before* the resume so that the frames which arrive when
 *   it comes back are already the right size.
 */
fun cameraDeliveryFor(
    surface: CameraSurface?,
    muted: Boolean,
    metered: Boolean,
): CameraDelivery = if (surface == null || muted) {
    CameraDelivery(enabled = false, quality = null)
} else {
    CameraDelivery(enabled = true, quality = cameraReceiveLayerFor(metered, surface))
}
